### ISSUE TITLE
HW watchpoint

### ISSUE THREAD

**Issue by fafok29** [nakst/gf#109]
https://github.com/nakst/gf/blob/2b048cfe6a1c095a5da74ab25f8d32fc09728f13/windows.cpp#L1803-L1806

Maybe the address should be cast to type because, with the current approach, we can miss writes to struct fields, If the type is unspecified gdb will use some default size (1,2,4,8), so if you have 16 bytes struct, second half of it won't be checked(given that gdb will use 8 byte by default)

But at least on my Intel Core i7-8550U if the data type is up to 32 bytes and we cast address to type gdb will create HW watchpoint, otherwise SW watchpoint will be created which is very slow. (on my machine beyond usable), but I think 32 bytes is usually good enough to be usable.

Also, 32 bytes size aligns with [x86 debug register](https://en.wikipedia.org/wiki/X86_debug_register) -> DR0 - DR3 up to 8 bytes each

The issue is that it is not clear how to detect this size without a hardcode, especially if it is not intel.

**Comment by nakst** [comment]
Maybe you could evaluate the `sizeof(%s)` expression first? I can't remember what I tried.

**Comment by nakst** [comment]
Maybe `watch -l` does the correct thing?

**Comment by fafok29** [comment]
> Maybe `watch -l` does the correct thing?

Yeah, it would be better with -l, but there is still a problem in that you don't know what is allowed size for the hardware watchpoint because GDB will silently create a software watchpoint if you exceed this size, 
Maybe just letting the user know that watching anything > 32-bit will be very slow, would be enough and 32-bit is a global limit not per watch. 
It wouldn't be hard to check this, since gdb tells the type of watchpoint.
